### PR TITLE
Modify Parallelization Strategy to Make it More General

### PR DESCRIPTION
**Pull request by zhenglongjiepheonix**
As per title, this PR tries a more general approach rather than relying purely on human heuristics, basically it uses the following steps to search a possible parallelization strategy for a transformer model

- Use dynamo for graph tracing so that we get the graph to operate on
- Decompose and functionalize the traced graph so that we get a smaller op set to work with
- Apply parallel axis analysis and do a constrained backtracking search on the whole graph to get a possible solution(not necessarily optimal)
- Replace ops the original traced graph with their parallelized version(Linear -> ColumnLinear/RowLinear)

And for the API design, we disable the support of passing custom modules and only focus on models in transformers because supporting custom models is not the priority for now.

### CODE REVIEW

File: optimum/fx/parallelization/decomp.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+                return super().run(*args, **kwargs)
+
+
+def decompose_and_functionalize(
+    graph_module: GraphModule,
+    decomposition_table: Dict[torch._ops.OperatorBase, Callable] = core_aten_decompositions(),
+    leaf_function_targets: List[Callable] = [F.scaled_dot_product_attention],
+) -> Callable:
+    """
+    API to decompose and funcitonalize a high-level graph module.
```
Comment (michaelbenayoun) [inline optimum/fx/parallelization/decomp.py:193]:
```suggestion
    API to decompose and functionalize a high-level graph module.
```

File: optimum/fx/parallelization/decomp.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+    Args:
+        graph_module (GraphModule):
+            The high-level graph module to be decomposed and functionalized.
+        decomposition_table (Dict[torch._ops.OperatorBase, Callable], defaults to `core_aten_decompostions()`):
+            The lookup table which maps high-level torch op to their equivalent low-level implementation.
+        leaf_function_targets (List[Callable], defaults to `[F.scaled_dot_product_attention]`):
```
Comment (michaelbenayoun) [inline optimum/fx/parallelization/decomp.py:200]:
```suggestion
        graph_module (`GraphModule`):
            The high-level graph module to be decomposed and functionalized.
        decomposition_table (`Dict[torch._ops.OperatorBase, Callable]`, defaults to `core_aten_decompostions()`):
            The lookup table which maps high-level torch op to their equivalent low-level implementation.
        leaf_function_targets (`List[Callable]`, defaults to `[F.scaled_dot_product_attention]`):
```

File: optimum/fx/parallelization/op_registry/op_handlers.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+                for op_type in op_types:
+                    self.mapping[op_type] = cls
+            else:
+                self.mapping[op_types] = cls
+            return cls
+
+        return wrapper
+
+    def is_supported(self, op_type) -> bool:
+        return op_type in self.mapping
```
Comment (michaelbenayoun) [inline optimum/fx/parallelization/op_registry/op_handlers.py:41]: What is this registry used for?

File: optimum/fx/parallelization/op_registry/op_handlers.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+        torch.ops.aten.bitwise_and.Scalar,
+        torch.ops.aten.bitwise_or.Scalar,
+        torch.ops.aten.bitwise_xor.Scalar,
+    ]
+)
+class UnaryOpParallelAxisPropagateHandler(OpParallelAxisPropagateHandler):
+    def propagate(self) -> bool:
+        arg = self.node.all_input_nodes[0]
+        axis = self.extract_axis(arg)
+        return [axis]
```
Comment (michaelbenayoun) [inline optimum/fx/parallelization/op_registry/op_handlers.py:130]: nit it's not returning a bool.
If I understand properly it returns the axis that is supposed to be parallel?

File: optimum/fx/parallelization/api.py [hunk 5]
```diff
@@ -43,17 +44,18 @@ def parallelize_backend(
 
 
 def parallelize_model(
-    model: Union[torch.nn.Module, str],
+    model: str,
     parallel_ctx: ParallelExecutionCtx,
     *model_args,
     **kwargs,
-):
+) -> Callable:
```
[the remark this answers is not in the document]
Comment (zhenglongjiepheonix) [reply optimum/fx/parallelization/api.py:51]: Not exactly，`torch.compile` returns a handle patched around the original module, and will trigger tracing and compilation when firstly called with data

File: optimum/fx/parallelization/op_registry/op_handlers.py (hunk 4 above)
Comment (zhenglongjiepheonix) [reply optimum/fx/parallelization/op_registry/op_handlers.py:130]: Yes, you are right, it looks up the axis of all the inputs and try deducting the axis of the output

File: optimum/fx/parallelization/op_registry/op_handlers.py (hunk 3 above)
Comment (zhenglongjiepheonix) [reply optimum/fx/parallelization/op_registry/op_handlers.py:41]: This is for registration of parallel axis policy of different aten ops

File: optimum/fx/parallelization/decomp.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    def create_node_(*args, **kwargs):
+        node = creat_node(*args, **kwargs)
+        node.meta["traced_from"] = traceback.get_current_meta()["from_node"]
+        return node
+
+    try:
+        Graph.create_node = create_node_
+        yield
+    finally:
+        Graph.create_node = creat_node
```
[the remark this answers is not in the document]
Comment (zhenglongjiepheonix) [reply optimum/fx/parallelization/decomp.py:60]: Yes, it records the original high-level node from which it comes from

File: optimum/fx/parallelization/op_registry/op_handlers.py (hunk 3 above)
Comment (michaelbenayoun) [reply optimum/fx/parallelization/op_registry/op_handlers.py:41]: Can you add a docstring to explain that please?

File: optimum/fx/parallelization/passes.py [hunk 7]
```diff
@@ -135,238 +134,151 @@ def clean_all(self, graph_module: GraphModule) -> None:
...
-            elif is_embedding(node):
-                # directly mark `nn.Embedding` layers
-                self.place_marker_per_node(node, {"axis": "vocab"})
 
-        return graph_module
+        nodes = list(graph.nodes)
 
+        def search(idx: int):
+            if idx == len(nodes):
+                return True
```
Comment (michaelbenayoun) [inline optimum/fx/parallelization/passes.py:201]:
```suggestion
        def search(idx: int) -> bool:
            return idx == len(nodes)     
```

Comment (zhenglongjiepheonix) [reply optimum/fx/parallelization/passes.py:201]: the `search` here actually is a backtracking search function entailing more logic following, so we can only return True if we have reached the very last op

File: optimum/fx/parallelization/op_registry/op_handlers.py [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+from torch.fx import Node
+
+from ..core import Config
+from ..utils import is_activation, is_embedding, is_linear
+
+
+class Registry:
+    """
+    Registry class handles registration of parallel axis propagation handlers of different aten ops, to support a new
+    aten op, you need to register the corresponding handler class by decorating it with `register` function.
```
Comment (michaelbenayoun) [inline optimum/fx/parallelization/op_registry/op_handlers.py:28]: nit:
 ```suggestion
    Registry class handles registration of parallel axis propagation handlers of different aten ops.
    To support a new aten op, you need to register the corresponding handler class by decorating 
    it with the `register` function.
```

**Review by michaelbenayoun — APPROVED** [review]
LGTM!

**Comment by zhenglongjiepheonix** [comment]
merge this for irrelevant failures
